cdc/sink/codec/canal.go

```diff
@@ -327,20 +335,94 @@ func (d *CanalEventBatchEncoder) EncodeCheckpointEvent(ts uint64) (*MQMessage, e
 
 // AppendRowChangedEvent implements the EventBatchEncoder interface
 func (d *CanalEventBatchEncoder) AppendRowChangedEvent(e *model.RowChangedEvent) (EncoderResult, error) {
+	d.txnCache.Append(nil, e)
+	return EncoderNoOperation, nil
+}
+
+// AppendResolvedEvent appends a resolved event to the encoder
+func (d *CanalEventBatchEncoder) AppendResolvedEvent(ts uint64) (EncoderResult, error) {
+	if ts < d.resolvedTs {
+		return EncoderNoOperation, nil
+	}
+	d.resolvedTs = ts
+	return EncoderNeedAsyncWrite, nil
+}
+
+// EncodeDDLEvent implements the EventBatchEncoder interface
+func (d *CanalEventBatchEncoder) EncodeDDLEvent(e *model.DDLEvent) (*MQMessage, error) {
+	canalMessageEncoder := newCanalMessageEncoder(d.forceHkPk)
+	return canalMessageEncoder.encodeDDLEvent(e)
+}
+
+// Build implements the EventBatchEncoder interface
+func (d *CanalEventBatchEncoder) Build() []*MQMessage {
+	resolvedTxns := d.txnCache.Resolved(d.resolvedTs)
+	if len(resolvedTxns) == 0 {
+		return nil
+	}
+	messages := make([]*MQMessage, 0, len(resolvedTxns))
+	for _, txns := range resolvedTxns {
+		for _, txn := range txns {
+			canalMessageEncoder := newCanalMessageEncoder(d.forceHkPk)
+			for _, row := range txn.Rows {
+				err := canalMessageEncoder.appendRowChangedEvent(row)
+				if err != nil {
+					log.Fatal("Error when append row change event", zap.Error(err))
+				}
+			}
+			messages = append(messages, canalMessageEncoder.build(txn.CommitTs))
+		}
+	}
+	sort.Slice(messages, func(i, j int) bool { return messages[i].Ts < messages[j].Ts })
+	return messages
+}
+
+// MixedBuild implements the EventBatchEncoder interface
+func (d *CanalEventBatchEncoder) MixedBuild(withVersion bool) []byte {
+	panic("Mixed Build only use for JsonEncoder")
+}
+
+//Size implements the EventBatchEncoder interface
+func (d *CanalEventBatchEncoder) Size() int {
+	// FIXME encoder with transaction support is hard to calculate the encoded buffer size
```

Comment:
After support transaction, knowing the encoded size becomes no need, because `Build()` method will only encode 
 event depend on the `ResolvedTs`, and call `Build()` without updating `ResolvedTs` won't reduce the encoders size.